File: python/aoc2022/days/day11.py
monkeys = []


def day11():
    with open('../input/input11.txt', 'r') as f:
        data = f.read().split('\n')

    for m in range(len(data) // 7):
        monkeys.append(Monkey(data[7 * m:(7 * m + 7)]))

    for _ in range(20):
        for monkey in monkeys:
            monkey.turn()

    monkey_business = [monkey.activity for monkey in monkeys]
    monkey_business.sort(reverse=True)
    print('Part 1:', monkey_business[0] * monkey_business[1])

    monkeys.clear()
    worry = 1
    for m in range(len(data) // 7):
        monkeys.append(Monkey(data[7 * m:(7 * m + 7)]))
        worry *= monkeys[m].test

    for _ in range(10000):
        for monkey in monkeys:
            monkey.turn(worry)

    monkey_business = [monkey.activity for monkey in monkeys]
    monkey_business.sort(reverse=True)
    print('Part 2:', monkey_business[0] * monkey_business[1])


class Monkey(object):

    def __init__(self, data):
        self.id = int(data[0].split(':')[0].split(' ')[1])
        self.items = [int(val) for val in data[1].split(':')[1].split(',')]
        op = data[2].split('new = old ')[1]
        self.oper = op.split(' ')[0]
        self.val = op.split(' ')[1]
        self.test = int(data[3].split('by')[1])
        self.t = int(data[4].split('monkey')[1])
        self.f = int(data[5].split('monkey')[1])
        self.activity = 0

    def __repr__(self):
        s = f"Monkey {self.id}, holding {str(self.items)}\nOperation : new = old {self.oper} {self.val}\n"
        s += f"If level divisible by {self.test}"
        s += f"\n- then pass to Monkey {self.t}, else:\n- pass to Monkey {self.f}\n"
        s += f"Its activity level is: {self.activity}\n"
        return s

    def turn(self, worry=None):
        for _ in range(len(self.items)):
            self.inspect()
            self.get_bored(worry)
            self.pass_item()

    def inspect(self):
        self.activity += 1
        if self.val == 'old':
            val = self.items[0]
        else:
            val = int(self.val)
        if self.oper == '+':
            self.items[0] = self.items[0] + val
        else:
            self.items[0] = self.items[0] * val

    def get_bored(self, worry=None):
        if worry is None:
            self.items[0] = self.items[0] // 3
        else:
            self.items[0] = self.items[0] % worry

    def pass_item(self):
        if self.items[0] % self.test == 0:
            monkeys[self.t].receive(self.items.pop(0))
        else:
            monkeys[self.f].receive(self.items.pop(0))

    def receive(self, item):
        self.items.append(item)

File: rust/src/aoc2022/day11.rs
use rdcl_aoc_helpers::input::WithReadLines;
use std::{fs::File, collections::VecDeque};

pub fn day11(path: &String) {
    println!("Day 11");
    let data:Vec<String> = File::open(path).read_lines::<String>(1).collect();

    let mut monkeys: MonkeyCircus = MonkeyCircus(Vec::new());
    for i in 0..((data.len()+1)/7 as usize) {
        monkeys.push(Monkey::new(data[7*i..(7*i+6)].to_vec()));
    }

    for _ in 0..20 {
        for i in 0..monkeys.0.len() {
            let nb_items = monkeys.0[i].items.len();
            for _ in 0..nb_items {
                let mut monkey = monkeys.0[i].clone();
                monkey.inspect();
                monkey.get_bored(0);
                let t = monkey.pass_to();
                let item = monkey.pass_item();
                monkeys.0[i] = monkey.clone();
                let mut friend = monkeys.0[t as usize].clone();
                friend.receive_item(item);
                monkeys.0[t as usize] = friend;
            }
        }
    }

    let mut monkey_business:Vec<u64> = monkeys.0.iter().map(|x| x.activity).collect();
    monkey_business.sort();
    monkey_business.reverse();
    println!("Part 1: {}", monkey_business[0] * monkey_business[1]);

    monkeys.0.clear();
    let mut worry = 1;

    for i in 0..((data.len()+1)/7 as usize) {
        let m = Monkey::new(data[7*i..(7*i+6)].to_vec());
        monkeys.push(m.clone());
        worry = worry * m.test;
    }

    for _ in 0..10000 {
        for i in 0..monkeys.0.len() {
            let nb_items = monkeys.0[i].items.len();
            for _ in 0..nb_items {
                let mut monkey = monkeys.0[i].clone();
                monkey.inspect();
                monkey.get_bored(worry);
                let t = monkey.pass_to();
                let item = monkey.pass_item();
                monkeys.0[i] = monkey.clone();
                let mut friend = monkeys.0[t as usize].clone();
                friend.receive_item(item);
                monkeys.0[t as usize] = friend;
            }
        }
    }

    monkey_business = monkeys.0.iter().map(|x| x.activity).collect();
    monkey_business.sort();
    monkey_business.reverse();
    println!("Part 2: {}", monkey_business[0] * monkey_business[1]);

}

#[derive(Debug)]
struct MonkeyCircus(Vec<Monkey>);

impl MonkeyCircus {
    fn push(&mut self, monkey: Monkey) {
        self.0.push(monkey);
    }    
}

#[derive(Debug, Clone)]
struct Monkey {
    items: VecDeque<u64>,
    oper: char,
    val: String,
    test: u64,
    t: u8,
    f:u8,
    activity: u64,
}

impl Monkey {
    fn new(data : Vec<String>) -> Monkey {
        let oper = data.iter().nth(2).unwrap().split("new = old ").last().unwrap().chars().nth(0).unwrap();
        let val = data.iter().nth(2).unwrap().split("new = old ").last().unwrap().split(' ').last().unwrap().to_string();
        let test =  data.iter().nth(3).unwrap().split("by ").last().unwrap().parse::<u64>().unwrap();
        let t =  data.iter().nth(4).unwrap().split("monkey ").last().unwrap().parse::<u8>().unwrap();
        let f =  data.iter().nth(5).unwrap().split("monkey ").last().unwrap().parse::<u8>().unwrap();
        let mut items:VecDeque<u64> = VecDeque::new();
        for item in data.iter().nth(1).unwrap().split(": ").last().unwrap().split(", "){
           items.push_back(item.parse::<u64>().unwrap());
        }
        Monkey { 
            items, 
            oper,
            val,
            test, 
            t, 
            f, 
            activity: 0 
        }
    }

    fn inspect(&mut self) {
        self.activity += 1;
        let val;
        if self.val == "old" {
            val = self.items[0];
        } else {
            val = self.val.parse::<u64>().unwrap();
        }
        if self.oper == '+' {
            self.items[0] = self.items[0] + val;
        } else {
            self.items[0] = self.items[0] * val;
        }
    }

    fn get_bored(& mut self, worry: u64) {
        if worry == 0{
            self.items[0] = self.items[0] / 3;
        }else{
            self.items[0] = self.items[0] % worry;
        }
    }

    fn pass_to(&self) -> u8 {
        if self.items[0] % self.test == 0{
            return self.t;
        }
        return self.f;
    }

    fn pass_item(&mut self) -> u64 {
        return self.items.pop_front().unwrap();
    }

    fn receive_item(&mut self, item:u64) {
        self.items.push_back(item);
    }
}